src/lib/operators/table_scan/abstract_single_column_table_scan_impl.cpp

```diff
@@ -0,0 +1,77 @@
+#include "abstract_single_column_table_scan_impl.hpp"
+
+#include <memory>
+#include <unordered_map>
+#include <utility>
+
+#include "resolve_type.hpp"
+#include "storage/chunk.hpp"
+#include "storage/dictionary_segment.hpp"
+#include "storage/reference_segment.hpp"
+#include "storage/split_pos_list_by_chunk_id.hpp"
+#include "storage/table.hpp"
+#include "storage/value_segment.hpp"
+
+namespace opossum {
+
+AbstractSingleColumnTableScanImpl::AbstractSingleColumnTableScanImpl(const std::shared_ptr<const Table>& in_table,
+                                                                     const ColumnID column_id,
+                                                                     const PredicateCondition predicate_condition)
+    : _in_table(in_table), _column_id(column_id), _predicate_condition(predicate_condition) {}
+
+std::shared_ptr<PosList> AbstractSingleColumnTableScanImpl::scan_chunk(const ChunkID chunk_id) const {
+  const auto& chunk = _in_table->get_chunk(chunk_id);
+  const auto& segment = chunk->get_segment(_column_id);
+
+  auto matches = std::make_shared<PosList>();
+
+  if (const auto& reference_segment = std::dynamic_pointer_cast<ReferenceSegment>(segment)) {
+    _scan_reference_segment(*reference_segment, chunk_id, *matches);
+  } else {
+    _on_scan(*segment, chunk_id, *matches, nullptr);
+  }
+
+  return matches;
+}
+
+void AbstractSingleColumnTableScanImpl::_scan_reference_segment(const ReferenceSegment& segment, const ChunkID chunk_id,
+                                                                PosList& matches) const {
+  const auto& pos_list = segment.pos_list();
+
+  if (pos_list->references_single_chunk() && !pos_list->empty()) {
+    // Fast path :)
+
+    const auto chunk = segment.referenced_table()->get_chunk(pos_list->common_chunk_id());
+    auto referenced_segment = chunk->get_segment(segment.referenced_column_id());
+
+    _on_scan(*referenced_segment, chunk_id, matches, pos_list);
+
+    return;
+  }
+
+  // Slow path - we are looking at multiple referenced chunks and need to split the pos list first
+  auto referenced_chunk_count = segment.referenced_table()->chunk_count();
+  auto chunk_offsets_by_chunk_id = split_pos_list_by_chunk_id(pos_list, referenced_chunk_count);
+
+  // Visit each referenced segment
+  for (auto referenced_chunk_id = ChunkID{0}; referenced_chunk_id < referenced_chunk_count; ++referenced_chunk_id) {
+    const auto& splitted_pos_list = chunk_offsets_by_chunk_id[referenced_chunk_id];
```

Comment:
`splitted_pos_list` is of type `SubPosList`, no, I'd name it so? Then the variable `sub_pos_list` below should be, to keep terminology consistent be the "position_filter", right?